### PR TITLE
First draft of Interactive token sale

### PR DESCRIPTION
**Pull request by joshuahannan**

### CODE REVIEW

File: InteractiveCrowdsale/truffle/contracts/InteractiveCrowdsaleLib.sol [hunk 1]
```diff
@@ -0,0 +1,371 @@
...
+
+    // subtract the bid from the balance of the owner
+    self.base.ownerBalance -= self.base.hasContributed[msg.sender];
+
+    // return bought tokens to the owners pool and remove tokens from the bidders pool
+    self.base.withdrawTokensMap[self.base.owner] += self.base.withdrawTokensMap[msg.sender];
+    self.base.withdrawTokensMap[msg.sender] = 0;
+
+    // iterate through the array of addresses at a certain valuation and remove the bidder's address who is withdrawing their bid
+    for (uint256 i = 0; i < self.valuationAddresses[self.personalValuations[msg.sender]].length; i++ ) {
```
Comment (mrsmkl) [inline InteractiveCrowdsale/truffle/contracts/InteractiveCrowdsaleLib.sol]: How to make sure this won't become too expensive?

Comment (Hackdom) [reply InteractiveCrowdsale/truffle/contracts/InteractiveCrowdsaleLib.sol]: That's one of the challenges to overcome. I mentioned in the Slack. We need to think about either limiting the iterations to a number and go through via batch calls from the sale owner or track the msg.gas variable and ensure the function can operate within limits somehow.

Comment (mrsmkl) [reply InteractiveCrowdsale/truffle/contracts/InteractiveCrowdsaleLib.sol]: Ok. Is it possible to replace this array with linked list?

Comment (joshuahannan) [reply InteractiveCrowdsale/truffle/contracts/InteractiveCrowdsaleLib.sol]: we could probably replace it. still depends on the gas costs of the different operations associated. I think figuring those out is going to be my next priority

**Pull request merged**